src/btree/bt_split.c

```diff
@@ -1539,45 +1571,57 @@ __split_multi_inmem_final(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *mul
 
         if (supd->ins == NULL) {
             slot = WT_ROW_SLOT(orig, supd->ripcip);
-            upd = orig->modify->mod_row_update[slot];
             orig->modify->mod_row_update[slot] = NULL;
-        } else {
-            upd = supd->ins->upd;
+        } else
             supd->ins->upd = NULL;
-        }
 
-        /*
-         * Free the onpage value and the older versions moved to the history store. We can't free
-         * the updates for in memory database and fixed length column store as they don't support
-         * the history sore.
-         */
+        /* Free the updates written to the data store and the history store. */
         if (supd->onpage_upd != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY) &&
           orig->type != WT_PAGE_COL_FIX) {
-            /*
-             * We have decided to restore this update chain so it must have newer updates than the
-             * onpage value on it.
-             */
-            WT_ASSERT(session, upd != NULL && upd != supd->onpage_upd && supd->onpage_upd != NULL);
-            /*
-             * Move the pointer to the position before the onpage value and free all the updates
-             * starting from the onpage value.
-             */
-            for (; upd->next != NULL && upd->next != supd->onpage_upd; upd = upd->next)
-                ;
-            WT_ASSERT(session, upd->next == supd->onpage_upd);
-            upd->next = NULL;
+            /* Track the freed size. */
+            for (upd = supd->onpage_upd; upd != NULL; upd = upd->next)
+                size += WT_UPDATE_MEMSIZE(upd);
```

Comment:
@keithbostic I'm not sure whether this is required. I'm tracking the memory size here because I don't want us to underestimate the amount of memory we freed.

Reply:
I now believe this is not necessary. We handle that in __wt_page_modify_clear in __split_multi.

Reply:
Reverted this change.